### PR TITLE
Route tooling redux

### PR DESCRIPTION
**Pull request by JamesNK**
This PR is https://github.com/dotnet/aspnetcore/pull/42597 but targeting .NET 8 and using Roslyn 4.4.0. The previous PR had to be reverted because source build changes couldn't make it into Roslyn 4.3.0.

### CODE REVIEW

**Comment by JamesNK** [comment]
@MichaelSimons The previous PR caused https://github.com/dotnet/aspnetcore/issues/42927. Is there a way to verify this is fixed without merging this PR?

**Comment by JamesNK** [comment]
Ah, ok, thanks for looking into this. I've updated the version of `Microsoft.CodeAnalysis.ExternalAccess.AspNetCore` to a newer one, removing some problematic dependencies.

> `prebuilt-usage.xml`

I see that the output XML file is in the build artifacts in CI 👍

**Comment by JamesNK** [comment]
In the new prebuilt-usage.xml all the problematic assemblies have been removed.

One exception is `Humanizer.Core` 2.14.1. It's used by `Microsoft.CodeAnalysis.CSharp.Features`, but it appears assembly is supported according to this issue I found in the source-build repo: https://github.com/dotnet/source-build/issues/2516

**Comment by MichaelSimons** [comment]
> One exception is Humanizer.Core which is still at version 2.14.1. It's used by Microsoft.CodeAnalysis.CSharp.Features, but it appears that Microsoft.CodeAnalysis.CSharp.Features is supported according to this issue I found in the source-build repo: https://github.com/dotnet/source-build/issues/2516. Does that mean it's ok?

This is a transitive dependency coming from other direct references.  Likely this will be resolved when built in the source-build tarball context because those dependencies will be force to use the source-built version of Humanizer.Core.  

Looking through the other diffs I did spot one direct reference which isn't included in source-build today.  

```
    <Usage Id="Microsoft.CodeAnalysis.Elfie" Version="1.0.0" File="src/artifacts/obj/Microsoft.AspNetCore.App.Analyzers/project.assets.json" />
    <Usage Id="Microsoft.CodeAnalysis.Elfie" Version="1.0.0" File="src/artifacts/obj/Microsoft.AspNetCore.App.CodeFixes/project.assets.json" />
```

The following questions need to be answered:

1. Is this dependency required for source-build?  Is the usage something that would be exercised in source-build?
2. If yes to the previous question, is the Elfie assembly something that can be brought into source-build?  What does its dependency graph look like?

**Comment by JamesNK** [comment]
I looked into `Microsoft.CodeAnalysis.Elfie` and it's a Windows-only thing. If the assembly is built using source build then a no-op implementation of the service is provided:

https://github.com/dotnet/roslyn/blob/5e8b6ee6b6bb294cfc67c768efad567d56122ef2/src/Features/Core/Portable/SymbolSearch/SourceBuild/SymbolSearchUpdateEngineFactory.cs#L4

The analyzer doesn't use symbol search so it seems fine.

**Comment by JamesNK** [comment]
@halter73 This is the second attempt of https://github.com/dotnet/aspnetcore/pull/42597. The source build issues from earlier should be sorted out.

The code is unchanged except some of the changes from https://github.com/dotnet/aspnetcore/pull/42910 slipped in. I deleted the big additions of the completion provider and its tests, but some shared code from that PR is included here. I don't think it is a big deal, just explaining why there are a couple of hundred extra lines of code in this PR compared to the old one 😄

File: AspNetCore.sln [hunk 1]
```diff
@@ -1745,6 +1746,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "WebTransportInteractiveSamp
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Templates.Blazor.WebAssembly.Auth.Tests", "src\ProjectTemplates\test\Templates.Blazor.WebAssembly.Auth.Tests\Templates.Blazor.WebAssembly.Auth.Tests.csproj", "{3A6FD623-F7F3-404B-8A39-CAFB40CA6A08}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "samples", "samples", "{890B5210-48EF-488F-93A2-F13BCB07C780}"
+EndProject
```
Comment (halter73) [inline AspNetCore.sln:1750]: Nit: Can we create a solution folder for `tests` too since we now have ones for both `src` and `samples`?

**Review by halter73 — APPROVED** [review]
> The code is unchanged except some of the changes from #42910 slipped in.

You didn't mean to add FrameworkParametersCompletionProvider, did you? Anyway. Once again, I'm super excited for the feature! 🎉 I'm looking forward to the autocompletion support too.

image

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/EmbeddedSyntax/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# About
+
+Embedded syntax code is copied from Roslyn:
+
+https://github.com/dotnet/roslyn/tree/b0de0c8e00ebf85db3c3884f2d81dfc3cb2d5a9d/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/Common
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/EmbeddedSyntax/README.md:5]: We're the only embedded syntax author. I've raised making the embedded extension point public to the Roslyn team already.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Microsoft.AspNetCore.App.Analyzers.csproj [hunk 3]
```diff
@@ -5,12 +5,12 @@
     <AddPublicApiAnalyzers>false</AddPublicApiAnalyzers>
     <TargetFramework>netstandard2.0</TargetFramework>
     <IncludeBuildOutput>false</IncludeBuildOutput>
-    <Nullable>Enable</Nullable>
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Microsoft.AspNetCore.App.Analyzers.csproj]: It is:

https://github.com/dotnet/aspnetcore/blob/92ff9a19481d90002959d24bcca09ef47e3386c8/eng/targets/CSharp.Common.targets#L18-L22

[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Microsoft.AspNetCore.App.Analyzers.csproj]: But, this project is netstandard2.0, so I don't think enabling nullable does anything anyway.

File: eng/Versions.props [hunk 4]
```diff
@@ -207,6 +207,7 @@
     -->
     <Analyzer_MicrosoftCodeAnalysisCSharpVersion>3.3.1</Analyzer_MicrosoftCodeAnalysisCSharpVersion>
     <Analyzer_MicrosoftCodeAnalysisCSharpWorkspacesVersion>3.3.1</Analyzer_MicrosoftCodeAnalysisCSharpWorkspacesVersion>
+    <MicrosoftCodeAnalysisExternalAccessAspNetCoreVersion>4.4.0-3.22472.13</MicrosoftCodeAnalysisExternalAccessAspNetCoreVersion>
```
Comment (RussKie) [inline eng/Versions.props:210]: /cc: @sharwell

**Pull request merged**